### DISCUSSION TITLE
Proposal: Implicit Generic Type Arguments

### DISCUSSION THREAD

**Discussion by ChristopherHaas** [dotnet/csharplang#4781] [Language Ideas]
The motivation for the feature is about reducing type parameters for types using nested generic types.

Suppose we have a type like this:

```
class Test<TCollection, TItem> where TCollection : ICollection<TItem>
{
	TCollection Property { get; }
 	TItem Find(int index);
}
```

The declaration of TITem in the type definition seems redundant, and this can cause some very long and complicated type names in more elaborate examples.

It would be awesome if the TItem did not need to be declared in type name but instead could be implicitly inferred from the constraint on TCollection. I.e., we could have something like:

```
class Test<TCollection> where TCollection : ICollection<TItem>
{
	TCollection Property { get; }
 	TItem Find(int index);
}
```

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/339?
